kyuubi-server/src/main/scala/org/apache/kyuubi/engine/flink/FlinkProcessBuilder.scala

```diff
@@ -50,88 +52,144 @@ class FlinkProcessBuilder(
 
   val flinkHome: String = getEngineHome(shortName)
 
+  val flinkExecutable: String = {
+    Paths.get(flinkHome, "bin", FLINK_EXEC_FILE).toFile.getCanonicalPath
+  }
+
   override protected def module: String = "kyuubi-flink-sql-engine"
 
   override protected def mainClass: String = "org.apache.kyuubi.engine.flink.FlinkSQLEngine"
 
   override def env: Map[String, String] = conf.getEnvs +
     (FLINK_PROXY_USER_KEY -> proxyUser)
 
+  override def clusterManager(): Option[String] = Some("yarn")
+
   override protected val commands: Array[String] = {
     KyuubiApplicationManager.tagApplication(engineRefId, shortName, clusterManager(), conf)
-    val buffer = new ArrayBuffer[String]()
-    buffer += executable
-
-    val memory = conf.get(ENGINE_FLINK_MEMORY)
-    buffer += s"-Xmx$memory"
-    val javaOptions = conf.get(ENGINE_FLINK_JAVA_OPTIONS)
-    if (javaOptions.isDefined) {
-      buffer += javaOptions.get
-    }
 
-    buffer += "-cp"
-    val classpathEntries = new java.util.LinkedHashSet[String]
-    // flink engine runtime jar
-    mainResource.foreach(classpathEntries.add)
-    // flink sql client jar
-    val flinkSqlClientPath = Paths.get(flinkHome)
-      .resolve("opt")
-      .toFile
-      .listFiles(new FilenameFilter {
-        override def accept(dir: File, name: String): Boolean = {
-          name.toLowerCase.startsWith("flink-sql-client")
+    // flink.execution.target are required in Kyuubi conf currently
+    val executionTarget = conf.getOption("flink.execution.target")
+    executionTarget match {
+      case Some("yarn-application") =>
+        val buffer = new ArrayBuffer[String]()
+        buffer += flinkExecutable
+        buffer += "run-application"
+
+        // write and set kyuubi configuration
+        val persistedConf = new Properties()
+        val ioTmpDir = System.getProperty("java.io.tmpdir")
+        val engineTmpDir = Paths.get(ioTmpDir, s"kyuubi-engine-$engineRefId")
+        Files.createDirectories(engineTmpDir)
+        val tmpKyuubiConfFile =
+          Paths.get(engineTmpDir.toAbsolutePath.toString, KYUUBI_CONF_FILE_NAME).toFile
+        val tmpKyuubiConf = tmpKyuubiConfFile.getCanonicalPath
+        // Scala 2.12 have ambiguous reference for properties#putAll with Java 11
+        // see https://github.com/scala/bug/issues/10418
+        conf.getAll.asJava.forEach((k, v) => persistedConf.put(k, v))
```

Comment:
is it possible to pass by cmd options? we do such a thing in `spark-submit --conf spark.kyuubi.xxx=yyy`, it benefits users to understand how flink engine works and easy to debug

Reply:
It's viable. The drawback is that the command line may get a bit lengthy. But I guess there's no chance it exceeds the max limit of the command line which is around 100,000.